### ISSUE TITLE
setHood and mapHood

### ISSUE THREAD

**Issue by jakebeal** [Protelis/Protelis#91]
In some cases, it would be better to have hood functions return a Java set or map object, rather than a Tuple.  I thus propose setHood and mapHood, which would have the following syntax and behavior:

* setHood is exactly like unionHood, except the output is a Java Set rather than a Tuple
* mapHood(x,y) is equivalent to unionHood([x, y]), except that the output is a Java Map with x as key and y as value.  One additional difference: if device A holds [x, y] and device B holds [x, y'], then mapHood will map x to precisely one of y or y' in an undetermined fashion.

**Comment by DanySK** [comment]
I believe this should be handled by operating over the returned tuple rather than through the definition of multiple hood operators.

**Comment by jakebeal** [comment]
I disagree for two reasons:
1. I believe that internally we *start* with a set, and so it should be pretty simple to keep it that way, rather than turning a set into a tuple and back again.
2. I'm finding myself writing a lot of painful glue code, which is push my local processing from Protelis to Java methods.

**Comment by DanySK** [comment]
I'm not convinced. Your first point is a mere implementative detail. Implementing those two functions would be easy whatever the implementation of a field looks like. They are just manipulations of a Tuple.
The second point does not justify two new builtin hoods either: the language must remain as small as possible, and what's needed should get added separately. The possible solutions that make sense to me are:
* Protelis code, to be included in the library shipped by default - that's my favourite option right now, and in perspective I believe we should drop also every whateverHood besides the pure `hood` itself
* Java code in form of static methods (i find this one not so nice)
* Transformation methods on Tuple (toMap(), toSet(), toList()...), to be chained to a classic unionHood (a bit more verbose)

I'd also like to hear @mviroli's opinion on the matter.

**Comment by jakebeal** [comment]
I'd be fine with turning everything except hood into Protelis code, as long as what we end up with is efficient.  What's problematic for me right now is turning everything into tuples just so that we can throw the tuples away again.  I think we should have the option of dealing with fields more directly.

**Comment by mviroli** [comment]
I concur that any only one hood operation has to be built-in, and others should be library.
On the other hand I see Protelis difficulties in handling "data structures", which should instead be smooth.
I think we should operate at the level of Protelis-Java integration: something I never found satisfactory, if you recall.
First a question: is there anything that prevents us from building a library Java class of static methods encapsulating powerful constructors and getters, e.g. for sets, and put it in the package so that one simply uses it in Protelis to work with sets just like it works with tuples?

**Comment by jakebeal** [comment]
I don't think that we even need to go that far.  I think that for the most part, we can simply use things like sets directly in Protelis.

**Comment by DanySK** [comment]
My current take on this is that we need to be able to easily use more data structures than tuples in Protelis, in general and not just when we reduce fields. Tuples were introduced at the time in order to have a data structure with no side effect. We probably need:
* Tuples (immutable lists)
* (immutable) Sets
* (immutable) Maps (or dicts, in python terminology)

Such a change would require to work both at the parser and interpreter level. If you agree on this, then I believe we should try to agree on the set of operations such entities should expose.

**Comment by mviroli** [comment]
I think that extending the language to directly manage some additional data types is something that goes too far. I will always need something else. For instance, the lesson we learn from modern functional languages (Scala) is that if I want to model a structured value ( e.g. a process with ID, name, expression, and creation time) like we will soon need to do, I do not use a tuple (which would lead to bad code), but rather I externally define a Java class with getters and factories, rigorously immutable perhaps, and I provide a once-and-for-all way of accessing them.

As I proposed above, I  do not understand why this is not already possible in Protelis.. though perhaps with improvable syntax.

**Comment by DanySK** [comment]
So, if I understand correctly, your proposal is to introduce something like: `DataStructure(arguments)` that produces what we want, along with an extensible library of data structures. Am I correct?

**Comment by mviroli** [comment]
More or less, yes. And in particular, I think you can already do this in Protelis, without changing syntax.
